### PR TITLE
Added a unit type

### LINKED ISSUE #293: Yul compilation failed: Call or assignment expected
Compiling this file:

```
contract FooProxy:
    pub def baz():
        ()
```

```
thread 'main' panicked at 'Yul compilation failed: {"errors":[{"component":"general","formattedMessage":"ParserError: Call or assignment expected.\n --> input.yul:1:168:\n  |\n1 |  ... e\" { code { function $$baz() { 0x0 } function avail() -> ptr { ptr := m ...\n  |                                         ^\n\n","message":"Call or assignment expected.","severity":"error","sourceLocation":{"end":168,"file":"input.yul","start":167},"type":"ParserError"},{"component":"general","formattedMessage":"ParserError: Expected keyword \"data\" or \"object\" or \"}\".\n --> input.yul:1:170:\n  |\n1 |  ...  { code { function $$baz() { 0x0 } function avail() -> ptr { ptr := mload(0x00 ...\n  |                                         ^^^^^^^^\n\n","message":"Expected keyword \"data\" or \"object\" or \"}\".","severity":"error","sourceLocation":{"end":177,"file":"input.yul","start":169},"type":"ParserError"}]}', compiler/src/lib.rs:48:27
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/panicking.rs:493:5
   1: std::panicking::begin_panic_fmt
             at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/panicking.rs:435:5
   2: fe_compiler::compile
   3: fe::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Aborted
```

On latest github master, built with

```
[profile.dev]
debug = true
panic = "abort"
```

for fuzzing.  Using https://github.com/agroce/afl-compiler-fuzzer

### LINKED ISSUE #407: Allow pass statement on contract level
### What is wrong?

I think that the `pass` statement should be allowed on the contract level. Currently it is only allowed on the function level.
The following code is illegal:

```
contract GuestBook:
    pass
```

### How can it be fixed

Start by teaching the parser to accept `pass` here and then follow the rabbit

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
### What was wrong?

closes #293 
closes #407 

We had two separate representations of empty tuples in the Yul output. One being `0x0` and the other being a function call with no return. This led to the issue discovered in #293.

### How was it fixed?

1.) Enabled empty struct definitions using `pass` (also did events and contracts).
2.) Lowered empty tuples to a struct type named `empty_tuple`.
3.) Removed all tuple mappings in the Yul codegen stage.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/docs/src/spec/index.md) if applicable
- [x] Add entry to the [release notes](https://github.com/ethereum/fe/blob/master/newsfragments/README.md) (may forgo for trivial changes)

- [x] Clean up commit history

### CODE REVIEW

**Comment by g-r-a-n-t** [comment]
decided Im going to hold off on this and wait for zero-sized structs so we can full implement tuple lowering

File: parser/src/grammar/types.rs [hunk 1]
```diff
@@ -27,10 +28,17 @@ pub fn parse_struct_def(par: &mut Parser) -> ParseResult<Node<ModuleStmt>> {
                 par.next()?;
                 break;
             }
+            Some(Pass) => {
+                parse_single_word_stmt(par)?;
+            }
             None => break,
-            Some(_) => {
+            Some(s) => {
```
Comment (g-r-a-n-t) [inline parser/src/grammar/types.rs:35]: remove debugging

File: parser/tests/cases/parse_ast.rs [hunk 2]
```diff
@@ -127,6 +127,9 @@ test_parse! { struct_def, types::parse_struct_def, r#"struct S:
   const z: u8
   pub const a: map<u8, foo>
 "# }
+test_parse! { empty_struct_def, types::parse_struct_def, r#"struct S:
```
Comment (g-r-a-n-t) [inline parser/tests/cases/parse_ast.rs:130]: add pass for events and contracts

File: test.fe [hunk 3]
```diff
@@ -0,0 +1,4 @@
+contract Ownable:
```
Comment (g-r-a-n-t) [inline test.fe:1]: delete this

File: compiler/src/lowering/mappers/functions.rs [hunk 4]
```diff
@@ -30,11 +35,34 @@ pub fn func_def(context: &Context, def: Node<fe::ContractStmt>) -> Node<fe::Cont
...
+            && !matches!(
+                lowered_body.last().unwrap().kind,
+                fe::FuncStmt::Return { .. }
+            )
+        {
+            lowered_body.push(
+                fe::FuncStmt::Return {
+                    value: Some(
+                        fe::Expr::Call {
+                            func: Box::new(fe::Expr::Name("empty_tuple".to_string()).into_node()),
```
Comment (g-r-a-n-t) [inline compiler/src/lowering/mappers/functions.rs:48]: cleanup

File: compiler/src/lowering/mappers/functions.rs [hunk 5]
```diff
@@ -30,11 +35,34 @@ pub fn func_def(context: &Context, def: Node<fe::ContractStmt>) -> Node<fe::Cont
...
+        }
+
         let lowered_kind = fe::ContractStmt::FuncDef {
             is_pub,
             name,
             args: lowered_args,
-            return_type: lowered_return_type,
+            return_type: Some(
+                lowered_return_type
+                    .unwrap_or(names::tuple_struct_type_desc(&Tuple::empty()).into_node()),
```
Comment (g-r-a-n-t) [inline compiler/src/lowering/mappers/functions.rs:64]: cleanup

File: analyzer/src/context.rs [hunk 6]
```diff
@@ -146,7 +146,7 @@ impl ExpressionAttributes {
 
     /// Return `true` if the type of the expression is an empty Tuple, otherwise
     /// `false`
-    pub fn is_empty_tuple(&self) -> bool {
```
Comment (g-r-a-n-t) [inline analyzer/src/context.rs:149]: update docs

File: analyzer/src/namespace/types.rs [hunk 7]
```diff
@@ -199,6 +199,10 @@ impl Struct {
         self.fields.is_empty()
     }
 
+    pub fn is_empty_tuple(&self) -> bool {
```
Comment (g-r-a-n-t) [inline analyzer/src/namespace/types.rs:202]: add docs

File: analyzer/src/namespace/types.rs [hunk 8]
```diff
@@ -323,7 +327,7 @@ impl Integer {
 
 impl Type {
     /// Returns true if the type is a tuple with 0 elements.
-    pub fn is_empty_tuple(&self) -> bool {
+    pub fn is_empty(&self) -> bool {
```
Comment (g-r-a-n-t) [inline analyzer/src/namespace/types.rs:330]: fix docs

File: analyzer/src/namespace/types.rs [hunk 9]
```diff
@@ -369,11 +373,19 @@ impl FixedSize {
     }
 
     pub fn is_empty_tuple(&self) -> bool {
```
Comment (g-r-a-n-t) [inline analyzer/src/namespace/types.rs:375]: add docs

File: analyzer/src/namespace/types.rs [hunk 10]
```diff
@@ -369,11 +373,19 @@ impl FixedSize {
...
-            return tuple.is_empty();
+        match self {
+            FixedSize::Tuple(tuple) => tuple.is_empty(),
+            FixedSize::Struct(_struct) => _struct.is_empty_tuple(),
+            _ => false,
         }
+    }
 
-        false
+    pub fn is_empty(&self) -> bool {
```
Comment (g-r-a-n-t) [inline analyzer/src/namespace/types.rs:383]: docs

File: analyzer/src/traversal/functions.rs [hunk 11]
```diff
@@ -31,6 +31,15 @@ pub fn func_def(
             .map(|arg| func_def_arg(Rc::clone(&function_scope), Rc::clone(&context), arg))
             .collect::<Result<Vec<_>, _>>()?;
 
+        if return_type.is_none() {
```
Comment (g-r-a-n-t) [inline analyzer/src/traversal/functions.rs:34]: explain

**Comment by sbillig** [comment]
@g-r-a-n-t It's maybe a bit late, but what if we introduce a Unit type (https://doc.rust-lang.org/std/primitive.unit.html) which just happens to resemble a tuple with no elements, and disallow empty tuples? We could use https://crates.io/crates/nonempty or https://crates.io/crates/vec1 to enforce the non-emptiness of the tuple elements vec.

That seems like a more aesthetically pleasing solution than lowering to an empty struct with a special name.

**Comment by g-r-a-n-t** [comment]
@sbillig Yeah, I think that makes sense. The `is_empty_tuple` functions are also something that I'm not a big fan of.

I'd be happy to try it out in this PR.

File: analyzer/src/namespace/types.rs [hunk 12]
```diff
@@ -812,6 +792,7 @@ impl SafeNames for FixedSize {
             FixedSize::String(string) => string.lower_snake(),
             FixedSize::Contract(contract) => contract.lower_snake(),
             FixedSize::Struct(val) => val.lower_snake(),
+            FixedSize::Unit => "unit".to_string(),
```
Comment (g-r-a-n-t) [inline analyzer/src/namespace/types.rs:795]: should panic

File: analyzer/src/namespace/types.rs [hunk 13]
```diff
@@ -948,10 +930,14 @@ pub fn type_desc(defs: &BTreeMap<String, Type>, typ: &fe::TypeDesc) -> Result<Ty
...
-                .iter()
-                .map(|typ| type_desc_fixed_size(defs, &typ.kind))
-                .collect::<Result<_, _>>()?,
+            items: Vec1::try_from_vec(
+                items
+                    .iter()
+                    .map(|typ| type_desc_fixed_size(defs, &typ.kind))
+                    .collect::<Result<_, _>>()?,
+            )
+            .expect("empty tuple"),
```
Comment (g-r-a-n-t) [inline analyzer/src/namespace/types.rs:939]: better message

File: analyzer/src/traversal/expressions.rs [hunk 14]
```diff
@@ -135,39 +137,32 @@ fn expr_tuple(
...
+                    .map(|attributes| attributes.typ)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
 
-            Ok(ExpressionAttributes::new(
-                Type::Tuple(tuple),
-                Location::Memory,
-            ))
+        let tuple = Tuple {
+            items: Vec1::try_from_vec(types_to_fixed_sizes(types)?).expect("emty tuple"),
```
Comment (g-r-a-n-t) [inline analyzer/src/traversal/expressions.rs:149]: better message

File: compiler/src/lowering/mappers/functions.rs [hunk 15]
```diff
@@ -14,9 +17,27 @@ pub fn func_def(context: &Context, def: Node<fe::ContractStmt>) -> Node<fe::Cont
...
-        let lowered_body = multiple_stmts(context, body);
+        let attributes = context.get_function(def.id).expect("missing attributes");
+
+        // The return type is lowered if it exists. If there is no return type, we set it to the empty tuple.
+        let lowered_return_type = return_type
+            .map(|return_type| types::type_desc(context, return_type))
+            .unwrap_or_else(|| names::fixed_size_type_desc(&FixedSize::Unit).into_node());
+
+        let lowered_body = {
+            let mut lowered_body = multiple_stmts(context, body);
```
Comment (g-r-a-n-t) [inline compiler/src/lowering/mappers/functions.rs:28]: update docs

File: compiler/src/yul/constructor.rs [hunk 16]
```diff
@@ -52,7 +52,8 @@ pub fn build_with_init(
 
         // add init function amd call it
         [init_func]
-        ([init_func_name]([decoded_params...]))
+        // init is given an empty tuple return during lowering, so we must pop it.
```
Comment (g-r-a-n-t) [inline compiler/src/yul/constructor.rs:55]: update all occurrences of "empty_tuple" or "empty tuple"

File: compiler/src/yul/runtime/abi_dispatcher.rs [hunk 17]
```diff
@@ -45,11 +45,10 @@ fn dispatch_arm(attributes: FunctionAttributes) -> yul::Case {
                 ([selection_with_return])
             }
         };
+    } else {
+        let selection = selection_as_pop_statement(&attributes.name, &attributes.param_types());
```
Comment (g-r-a-n-t) [inline compiler/src/yul/runtime/abi_dispatcher.rs:49]: explain with docs

File: compiler/tests/cases/lowering/fixtures/return_empty_tuple_lowered.fe [hunk 18]
```diff
@@ -0,0 +1,34 @@
+contract Foo:
```
Comment (g-r-a-n-t) [inline compiler/tests/cases/lowering/fixtures/return_empty_tuple_lowered.fe:1]: update the names of these files

**Comment by g-r-a-n-t** [comment]
Added a unit type @sbillig, this is ready for a review.

**Review by sbillig — APPROVED** [review]
:+1:

File: newsfragments/406.internal.md [hunk 19]
```diff
@@ -0,0 +1 @@
+Empty tuples are lowered to an empty struct.
```
Comment (sbillig) [inline newsfragments/406.internal.md:1]: This should be updated

**Review by sbillig — CHANGES_REQUESTED** [review]

**Review by sbillig — APPROVED** [review]

**Pull request merged**